pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix

```
@@ -0,0 +1,177 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+{ lib
+, cargo-pgx
+, pkg-config
+, runCommand
+, rustPlatform
+, rustc
+, rustfmt
+, stdenv
+, Security
+}:
+
+#
+# The idea behind: Use it mostly like rustPlatform.buildRustPackage and so
+# we hand most of the arguments down.
+#
+# The only additional argument is `postgresql`.
+# You need to assign the targeted postgresql package to this argument.
+#
+
+{ buildAndTestSubdir ? null
+, buildType ? "release"
+, buildFeatures ? [ ]
+, cargoBuildFlags ? [ ]
+, postgresql
+, ...
+} @ args:
+
+assert lib.asserts.assertMsg ((args.installPhase or "") == "")
+  "buildPgxExtensions overwrites the installPhase, so providing one is useless";
+assert lib.asserts.assertMsg ((args.buildPhase or "") == "")
+  "buildPgxExtensions overwrites the buildPhase, so providing one is useless";
+
+let
+  maybeDebugFlag = if buildType != "release" then "--debug" else "";
+  maybeEnterBuildAndTestSubdir =
+    if buildAndTestSubdir != null then ''
+      export CARGO_TARGET_DIR="$(pwd)/target"
+      pushd ${buildAndTestSubdir}
+    '' else "";
+  maybeLeaveBuildAndTestSubdir = if buildAndTestSubdir != null then "popd" else "";
+
+  pgxPostgresMajor = builtins.head (lib.splitString "." postgresql.version);
+  preBuildAndTest = ''
+    export PGX_HOME=$(mktemp -d)
+    mkdir -p $PGX_HOME/${pgxPostgresMajor}
+
+    cp -r -L ${postgresql}/. $PGX_HOME/${pgxPostgresMajor}/
+    chmod -R ugo+w $PGX_HOME/${pgxPostgresMajor}
+    cp -r -L ${postgresql.lib}/lib/. $PGX_HOME/${pgxPostgresMajor}/lib/
+
+    ${cargo-pgx}/bin/cargo-pgx pgx init \
+      --pg${pgxPostgresMajor} $PGX_HOME/${pgxPostgresMajor}/bin/pg_config
+
+    # This is primarily for Mac or other Nix systems that don't use the nixbld user.
+    export USER=$(whoami)
+    export PGDATA=$PGX_HOME/data-${pgxPostgresMajor}/
+    export NIX_PGLIBDIR=$PGX_HOME/${pgxPostgresMajor}/lib
+
+    echo "unix_socket_directories = '$(mktemp -d)'" > $PGDATA/postgresql.conf
+    ${postgresql}/bin/pg_ctl start
+    ${postgresql}/bin/createuser -h localhost --superuser --createdb $USER || true
+    ${postgresql}/bin/pg_ctl stop
+  '';
+
+  argsForBuildRustPackage = builtins.removeAttrs args [ "postgresql" ];
+
+  # so we don't accidentally `(rustPlatform.buildRustPackage argsForBuildRustPackage) // { ... }` because
+  # we forgot parentheses
+  finalArgs = argsForBuildRustPackage // {
+    buildInputs = (args.buildInputs or [ ]) ++ [ postgresql ] ++ lib.optionals stdenv.isDarwin [ Security ];
+
+    nativeBuildInputs = (args.nativeBuildInputs or [ ]) ++ [
+      cargo-pgx
+      rustc
+      rustfmt
+      pkg-config
+      rustPlatform.bindgenHook
+    ];
+
+    postPatch = ''
+      ${args.postPatch or ""}
+      patchShebangs .
+    '';
```

```suggestion
    postPatch = args.postPatch or "" + ''
      patchShebangs .
    '';
```
Also what files need to be patched? Patching all files is rather slow.

---

I copied and pasted this from cargo-pgx example code. I'll investigate, wether its really needed like that. Otherwise, it could simply be erased I assume and considered a responsibility of the caller.

---

I think it is not necessary at all and if someone needs it, he can set `postPatch` himself. I will remove it.